### ISSUE TITLE
Title: torch.compile() produces incorrect outputs for transformer model with causal attention

### ISSUE THREAD

**Issue by zhouxiaoyaozzz** [pytorch/pytorch#158226]
### 🐛 Describe the bug

When torch.compile() is used on a transformer model with causal relationships, the compiled model produces significantly different outputs compared to the uncompiled version, and even the first two elements of the model output have opposite positive and negative values compared to the uncompiled execution
code:
'''
import os
import torch
import torch.nn as nn
import torch.nn.functional as F
class CausalAttention(nn.Module):
    def __init__(self, embed_size, heads):
        super().__init__()
        self.embed_size = embed_size
        self.heads = heads
        self.head_dim = embed_size // heads

        assert self.head_dim * heads == embed_size, "Embed size needs to be divisible by heads"

        self.values = nn.Linear(self.head_dim, self.head_dim, bias=False)
        self.keys = nn.Linear(self.head_dim, self.head_dim, bias=False)
        self.queries = nn.Linear(self.head_dim, self.head_dim, bias=False)
        self.fc_out = nn.Linear(heads * self.head_dim, embed_size)

    def forward(self, values, keys, query, mask):
        N = query.shape[0]
        value_len, key_len, query_len = values.shape[1], keys.shape[1], query.shape[1]

        values = values.reshape(N, value_len, self.heads, self.head_dim)
        keys = keys.reshape(N, key_len, self.heads, self.head_dim)
        queries = query.reshape(N, query_len, self.heads, self.head_dim)

        values = self.values(values)
        keys = self.keys(keys)
        queries = self.queries(queries)

        energy = torch.einsum("nqhd,nkhd->nhqk", [queries, keys])

        if mask is not None:
            energy = energy.masked_fill(mask == 0, float("-1e20"))

        attention = torch.softmax(energy / (self.embed_size ** (1 / 2)), dim=3)

        out = torch.einsum("nhql,nlhd->nqhd", [attention, values]).reshape(
            N, query_len, self.heads * self.head_dim
        )

        out = self.fc_out(out)
        return out

class CausalAttentionDNN(nn.Module):
    def __init__(self, input_size, embed_size, num_layers, heads, forward_expansion, output_size, dropout, max_length):
        super(CausalAttentionDNN, self).__init__()
        self.embed_size = embed_size
        self.word_embedding = nn.Embedding(input_size, embed_size)
        self.position_embedding = nn.Embedding(max_length, embed_size)

        self.layers = nn.ModuleList(
            [
                CausalAttentionBlock(
                    embed_size,
                    heads,
                    forward_expansion,
                    dropout,
                )
                for _ in range(num_layers)
            ]
        )

        self.fc = nn.Linear(embed_size, output_size)
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, mask):
        N, seq_length = x.shape
        positions = torch.arange(0, seq_length).expand(N, seq_length).to(x.device)

        out = self.dropout(self.word_embedding(x) + self.position_embedding(positions))

        for layer in self.layers:
            out = layer(out, out, out, mask)

        out = self.fc(out)
        return out

class CausalAttentionBlock(nn.Module):
    def __init__(self, embed_size, heads, forward_expansion, dropout):
        super().__init__()
        self.attention = CausalAttention(embed_size, heads)
        self.norm1 = nn.LayerNorm(embed_size)
        self.norm2 = nn.LayerNorm(embed_size)

        self.feed_forward = nn.Sequential(
            nn.Linear(embed_size, forward_expansion * embed_size),
            nn.ReLU(),
            nn.Linear(forward_expansion * embed_size, embed_size),
        )

        self.dropout = nn.Dropout(dropout)

    def forward(self, value, key, query, mask):
        attention = self.attention(value, key, query, mask)
        x = query + self.dropout(self.norm1(attention))
        forward = self.feed_forward(x)
        out = x + self.dropout(self.norm2(forward))
        return out

def create_causal_mask(size):
    mask = torch.tril(torch.ones(size, size))
    return mask.unsqueeze(0).unsqueeze(0)

if __name__ == "__main__":

    input_size = 10000
    embed_size = 256
    num_layers = 6
    heads = 8
    forward_expansion = 4
    output_size = 10000
    dropout = 0.1
    max_length = 512

    model = CausalAttentionDNN(
        input_size, embed_size, num_layers, heads,
        forward_expansion, output_size, dropout, max_length
    )

    x = torch.randint(0, input_size, (32, 100))
    mask = create_causal_mask(x.shape[1])

    print("Running without torch.compile()...")
    output_normal = model(x, mask)
    print(f"Output shape (normal): {output_normal.shape}")
    print(f"First 5 elements of first output (normal): {output_normal[0, 0, :5]}")

    print("\nRunning with torch.compile()...")
    compiled_model = torch.compile(model)
    output_compiled = compiled_model(x, mask)
    print(f"Output shape (compiled): {output_compiled.shape}")
    print(f"First 5 elements of first output (compiled): {output_compiled[0, 0, :5]}")

    print("\nComparison:")
    print(f"Max absolute difference: {torch.max(torch.abs(output_normal - output_compiled))}")
    print(f"Mean absolute difference: {torch.mean(torch.abs(output_normal - output_compiled))}")
    print(f"Are all outputs close (1e-5 tolerance)? {torch.allclose(output_normal, output_compiled, atol=1e-5)}")

'''
output:
'''
Running without torch.compile()...
Output shape (normal): torch.Size([32, 100, 10000])
First 5 elements of first output (normal): tensor([ 0.4877, -0.6702,  1.1478, -2.2490, -1.5502], grad_fn=<SliceBackward0>)

Running with torch.compile()...
Output shape (compiled): torch.Size([32, 100, 10000])
First 5 elements of first output (compiled): tensor([-1.5253,  1.8060,  2.6570, -2.0276, -1.3364], grad_fn=<SliceBackward0>)

Comparison:
Max absolute difference: 8.375188827514648
Mean absolute difference: 1.1046061515808105
Are all outputs close (1e-5 tolerance)? False
'''
### Versions

Collecting environment information...
PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 4.0.3
Libc version: glibc-2.39

Python version: 3.9.7 (default, Apr 24 2025, 12:37:02)  [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.11.0-29-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
架构：                                x86_64
CPU 运行模式：                        32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
字节序：                              Little Endian
CPU:                                  20
在线 CPU 列表：                       0-19
厂商 ID：                             GenuineIntel
型号名称：                            Intel(R) Core(TM) i5-14500
CPU 系列：                            6
型号：                                191
每个核的线程数：                      2
每个座的核数：                        14
座：                                  1
步进：                                2
CPU(s) scaling MHz:                   64%
CPU 最大 MHz：                        5000.0000
CPU 最小 MHz：                        800.0000
BogoMIPS：                            5222.40
标记：                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
虚拟化：                              VT-x
L1d 缓存：                            544 KiB (14 instances)
L1i 缓存：                            704 KiB (14 instances)
L2 缓存：                             11.5 MiB (8 instances)
L3 缓存：                             24 MiB (1 instance)
NUMA 节点：                           1
NUMA 节点0 CPU：                      0-19
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] nvidia-cublas-cu11==11.10.3.66
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu11==11.7.101
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu11==11.7.99
[pip3] nvidia-cuda-nvrtc-cu12==12.6.77
[pip3] nvidia-cuda-runtime-cu11==11.7.99
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu11==8.5.0.96
[pip3] nvidia-cudnn-cu12==9.5.1.17
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu11==10.2.10.91
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu11==11.4.0.1
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu11==11.7.4.91
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu11==2.14.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu11==11.7.91
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.7.1
[pip3] torchdiffeq==0.2.5
[pip3] torchvision==0.22.1
[pip3] triton==3.3.1
[conda] Could not collect

cc @chauhang @penguinwu

**Comment by desertfire** [comment]
This is not a problem with torch.compile. In your model, there are dropout ops which are non-deterministic. If you set your model to .eval(), the results will be the same. Or you need to reset the random generator state between the two runs.
